File: app/src/main/java/com/solverlabs/worldcraft/activity/WorldCraftActivity.java
package com.solverlabs.worldcraft.activity;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.util.Log;

import com.solverlabs.droid.rugl.Game;
import com.solverlabs.droid.rugl.GameActivity;
import com.solverlabs.droid.rugl.input.Touch;
import com.solverlabs.droid.rugl.res.ResourceLoader;
import com.solverlabs.droid.rugl.texture.TextureFactory;
import com.solverlabs.droid.rugl.util.geom.Vector2f;
import com.solverlabs.droid.rugl.util.geom.Vector3f;
import com.solverlabs.worldcraft.BlockView;
import com.solverlabs.worldcraft.GameMode;
import com.solverlabs.worldcraft.MyApplication;
import com.solverlabs.worldcraft.Persistence;
import com.solverlabs.worldcraft.Player;
import com.solverlabs.worldcraft.SoundManager;
import com.solverlabs.worldcraft.World;
import com.solverlabs.worldcraft.dialog.DeathMenuDialog;
import com.solverlabs.worldcraft.factories.BlockFactory;
import com.solverlabs.worldcraft.multiplayer.Multiplayer;
import com.solverlabs.worldcraft.multiplayer.dialogs.ReportAbuseDialog;
import com.solverlabs.worldcraft.nbt.RegionFileCache;
import com.solverlabs.worldcraft.nbt.Tag;
import com.solverlabs.worldcraft.nbt.TagLoader;
import com.solverlabs.worldcraft.ui.CustomProgressDialog;
import com.solverlabs.worldcraft.util.GameTime;
import com.solverlabs.worldcraft.util.WorldGenerator;

import org.apache.commons.compress.archivers.cpio.CpioConstants;

import java.io.File;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class WorldCraftActivity extends GameActivity {
    protected MyApplication application;
    private BlockView bw;
    private DeathMenuDialog deathMenuDialog;
    private boolean isResumingGame = false;
    private ProgressDialog loadingProgressDialog;
    private ProgressDialog resumeDialog;
    private World world;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.application = (MyApplication) getApplicationContext();
        getWindow().addFlags(CpioConstants.C_IWUSR);
        String worldFileName = getIntent().getExtras().getString("world");
        boolean isNewGame = getIntent().getExtras().getBoolean("isNewGame");
        WorldGenerator.Mode gameMode = (WorldGenerator.Mode) getIntent().getExtras().get("gameMode");
        Touch.resetTouch();
        initSoundManager();
        Log.e(Game.RUGL_TAG, "loading " + worldFileName);
        File dir = new File(worldFileName);
        showProgressDialog();
        TagLoader tl = new TagLoad(new File(dir, World.LEVEL_DAT_FILE_NAME), dir, isNewGame, gameMode);
        tl.selfCompleting = true;
        ResourceLoader.load(tl);
    }

    private void initSoundManager() {
        SoundManager.initSounds(this);
        SoundManager.loadSounds();
    }

    @Override
    public void onPause() {
        clearReferences();
        if (world != null) {
            world.save();
        }
        isResumingGame = true;
        if (loadingProgressDialog != null) {
            loadingProgressDialog.dismiss();
        }
        if (GameMode.isMultiplayerMode()) {
            Multiplayer.instance.shutdown();
            finish();
        }
        super.onPause();
    }

    private void clearReferences() {
        if (application.isCurrentActivity(this)) {
            application.setCurrentActivity(null);
        }
    }

    @Override
    public void onResume() {
        application.setCurrentActivity(this);
        if (isResumingGame) {
            isResumingGame = false;
            if (GameMode.isMultiplayerMode()) {
                super.finish();
                return;
            }
            showResumeDialog();
        }
        super.onResume();
    }

    @Override
    public void onDestroy() {
        TextureFactory.removeListener();
        if (bw != null) {
            bw.destroyWorld();
            bw = null;
            TextureFactory.deleteAllTextures();
        }
        RegionFileCache.clear();
        System.runFinalization();
        Runtime.getRuntime().gc();
        super.onDestroy();
    }

    @Override
    public void finish() {
        if (this.world != null && world.isNewGame) {
            deleteSave(world.dir.getAbsolutePath());
        }
        super.finish();
    }

    private void showResumeDialog() {
        if (resumeDialog == null) {
            resumeDialog = new ProgressDialog(this);
            resumeDialog.setTitle("Please wait");
            resumeDialog.setMessage("Resuming");
        }
        runOnUiThread(() -> resumeDialog.show());
    }

    private void showProgressDialog() {
        if (loadingProgressDialog == null) {
            loadingProgressDialog = new CustomProgressDialog(this) {
                @Override
                public void onBackPressed() {
                    if (world != null) {
                        world.setCancel(true);
                    }
                    finish();
                }

                @Override
                public void buttonClick() {
                    if (world != null) {
                        world.setCancel(true);
                    }
                    finish();
                }
            };
        }
        loadingProgressDialog.show();
    }

    public void dismissProgresDialog() {
        loadingProgressDialog.dismiss();
    }

    public void dismissResumeDialog() {
        resumeDialog.dismiss();
    }

    public void dismissAllLoadingDialogs() {
        if (loadingProgressDialog != null && loadingProgressDialog.isShowing()) {
            loadingProgressDialog.dismiss();
        }
        if (resumeDialog != null && resumeDialog.isShowing()) {
            resumeDialog.dismiss();
        }
    }

    private void deleteSave(String absolutePath) {
        File dir = new File(absolutePath);
        String[] list = dir.list();
        if (list != null) {
            for (String str : list) {
                File file = new File(dir, str);
                if (file.isFile()) {
                    Log.d("del_file", "  " + file.delete());
                }
                if (file.isDirectory()) {
                    deleteSave(file.getAbsolutePath());
                }
            }
            if (dir.isDirectory() && list.length == 0) {
                Log.d("del_dir", "  " + dir.delete());
            }
        }
    }

    public class TagLoad extends TagLoader {
        final File dir;
        final WorldGenerator.Mode gameMode;
        final boolean isNewGame;

        TagLoad(File dat, File file, boolean z, WorldGenerator.Mode mode) {
            super(dat);
            this.dir = file;
            this.isNewGame = z;
            this.gameMode = mode;
        }

        @Override
        public void complete() {
            runOnUiThread(new TagLoaderRunnable());
        }

        public class TagLoaderRunnable implements Runnable {
            @Override
            public void run() {
                int mapType;
                if (resource == null) {
                    showToast("Could not load world level.dat\n" + exception.getClass().getSimpleName() + ":" + exception.getMessage(), true);
                    finish();
                    return;
                }
                try {
                    Tag time = resource.findTagByName(WorldGenerator.LAST_PLAYED);
                    GameTime.initTime((Long) time.getValue());
                    Tag playerTag = resource.findTagByName("Player");
                    Tag pos = playerTag.findTagByName("Pos");
                    Tag[] tl = (Tag[]) pos.getValue();
                    Vector3f p = new Vector3f();
                    p.x = ((Double) tl[0].getValue()).floatValue();
                    p.y = ((Double) tl[1].getValue()).floatValue();
                    p.z = ((Double) tl[2].getValue()).floatValue();
                    Tag rotaionTag = playerTag.findTagByName("Rotation");
                    Vector2f rotation = new Vector2f();
                    if (rotaionTag != null) {
                        Tag[] tl2 = (Tag[]) rotaionTag.getValue();
                        rotation.x = (Float) tl2[0].getValue();
                        rotation.y = (Float) tl2[1].getValue();
                        if (rotation.y > 1.5707964f || rotation.y < -1.5707964f) {
                            rotation.y = 0.0f;
                        }
                    }
                    Tag mapTypeTag = resource.findTagByName(WorldGenerator.MAP_TYPE);
                    if (mapTypeTag != null) {
                        mapType = (Integer) mapTypeTag.getValue();
                    } else {
                        mapType = -1;
                    }
                    world = new World(dir, p, resource) {
                        @Override
                        public boolean isLoadingDialogVisible() {
                            return loadingProgressDialog != null && loadingProgressDialog.isShowing();
                        }

                        @Override
                        public void incLoadingProgressStatus(int diff) {
                            loadingProgressDialog.incrementProgressBy(diff);
                        }

                        @Override
                        public void setLoadingProgressStatus(int progress, int max) {
                            loadingProgressDialog.setMax(max);
                            loadingProgressDialog.setProgress(progress);
                        }

                        @Override
                        public void showGameMenu() {
                            runOnUiThread(WorldCraftActivity.this::showGameMenuDialog);
                        }

                        @Override
                        public void dismissLoadingDialog() {
                            if (loadingProgressDialog != null) {
                                loadingProgressDialog.dismiss();
                            }
                        }

                        @Override
                        public void dismissLoadingDialogAndWait() {
                            if (loadingProgressDialog != null && loadingProgressDialog.isShowing()) {
                                final CountDownLatch latch = new CountDownLatch(1);
                                loadingProgressDialog.setOnDismissListener(dialog -> latch.countDown());
                                loadingProgressDialog.dismiss();
                                try {
                                    latch.await(2L, TimeUnit.SECONDS);
                                } catch (InterruptedException e) {
                                    e.printStackTrace();
                                }
                            }
                        }

                        @Override
                        public void showChat() {
                            runOnUiThread(WorldCraftActivity.this::showChatDialog);
                        }

                        @Override
                        public void showReportAbuse() {
                            runOnUiThread(() -> {
                                ReportAbuseDialog reportAbuseDialog = new ReportAbuseDialog(WorldCraftActivity.this);
                                reportAbuseDialog.show();
                            });
                        }

                        @Override
                        public void showDeathMenu(final Player player) {
                            runOnUiThread(() -> {
                                if (deathMenuDialog == null || !deathMenuDialog.isVisible()) {
                                    deathMenuDialog = new DeathMenuDialog(WorldCraftActivity.this, player);
                                    deathMenuDialog.show();
                                }
                            });
                        }
                    };
                    world.setNewGame(isNewGame);
                    world.setMapType(mapType);
                    ((CustomProgressDialog) loadingProgressDialog).updateMax(World.getLoadingLimit(isNewGame));
                    bw = new BlockView(world);
                    if (GameMode.isMultiplayerMode()) {
                        Multiplayer.instance.blockView = bw;
                    }
                    bw.setCamRotation(rotation);
                    bw.cam.invert = Persistence.getInstance().isInvertY();
                    GameMode.setGameMode((GameMode.isMultiplayerMode() || !WorldGenerator.Mode.SURVIVAL.equals(gameMode)) ? 1 : 0);
                    if (GameMode.isSurvivalMode() && !GameMode.isMultiplayerMode()) {
                        world.initSunLight();
                    }
                    initFog();
                    Game game = new Game(WorldCraftActivity.this, null, bw);
                    start(game);
                } catch (Exception e) {
                    showToast("Problem parsing level.dat - Maybe a corrupt file?", true);
                    Log.e(Game.RUGL_TAG, "Level.dat corrupted?", e);
                    finish();
                }
            }

            private void initFog() {
                float fogDistance = Persistence.getInstance().getFogDistance();
                if (fogDistance < 0.0f) {
                    fogDistance = 0.0f;
                }
                BlockFactory.state.fog.start = fogDistance;
                BlockFactory.state.fog.end = 10.0f + fogDistance;
            }
        }
    }
}

File: app/src/main/java/com/solverlabs/worldcraft/activity/WorldCraftActivity.kt
package com.solverlabs.worldcraft.activity

import android.app.ProgressDialog
import android.os.Bundle
import android.util.Log
import com.solverlabs.droid.rugl.Game
import com.solverlabs.droid.rugl.GameActivity
import com.solverlabs.droid.rugl.input.Touch
import com.solverlabs.droid.rugl.res.ResourceLoader
import com.solverlabs.droid.rugl.texture.TextureFactory
import com.solverlabs.droid.rugl.util.geom.Vector2f
import com.solverlabs.droid.rugl.util.geom.Vector3f
import com.solverlabs.worldcraft.*
import com.solverlabs.worldcraft.GameMode.isMultiplayerMode
import com.solverlabs.worldcraft.GameMode.isSurvivalMode
import com.solverlabs.worldcraft.GameMode.setGameMode
import com.solverlabs.worldcraft.dialog.DeathMenuDialog
import com.solverlabs.worldcraft.factories.BlockFactory
import com.solverlabs.worldcraft.multiplayer.Multiplayer
import com.solverlabs.worldcraft.multiplayer.dialogs.ReportAbuseDialog
import com.solverlabs.worldcraft.nbt.RegionFileCache
import com.solverlabs.worldcraft.nbt.Tag
import com.solverlabs.worldcraft.nbt.TagLoader
import com.solverlabs.worldcraft.ui.CustomProgressDialog
import com.solverlabs.worldcraft.util.GameTime
import com.solverlabs.worldcraft.util.WorldGenerator
import org.apache.commons.compress.archivers.cpio.CpioConstants
import java.io.File
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class WorldCraftActivity : GameActivity() {
    private lateinit var application: MyApplication
    private var bw: BlockView? = null
    private var deathMenuDialog: DeathMenuDialog? = null
    private var isResumingGame = false
    private var loadingProgressDialog: ProgressDialog? = null
    private var resumeDialog: ProgressDialog? = null
    private var world: World? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        application = applicationContext as MyApplication
        window.addFlags(CpioConstants.C_IWUSR)

        intent.extras?.let { extras ->
            val isNewGame = extras.getBoolean("isNewGame")
            val gameMode = extras["gameMode"] as WorldGenerator.Mode?

            Touch.resetTouch()
            initSoundManager()

            extras.getString("world")?.let { worldName ->
                showProgressDialog()
                val dir = File(worldName)
                TagLoad(File(dir, World.LEVEL_DAT_FILE_NAME), dir, isNewGame, gameMode).apply {
                    selfCompleting = true
                }.also { ResourceLoader.load(it) }
            }
        }
    }

    private fun initSoundManager() {
        SoundManager.initSounds(this)
        SoundManager.loadSounds()
    }

    override fun onPause() {
        clearReferences()
        world?.save()
        isResumingGame = true
        if (loadingProgressDialog != null) {
            loadingProgressDialog!!.dismiss()
        }
        if (isMultiplayerMode) {
            Multiplayer.instance.shutdown()
            finish()
        }
        super.onPause()
    }

    private fun clearReferences() {
        if (application.isCurrentActivity(this)) {
            application.currentActivity = null
        }
    }

    override fun onResume() {
        application.currentActivity = this
        if (isResumingGame) {
            isResumingGame = false
            if (isMultiplayerMode) {
                super.finish()
                return
            }
            showResumeDialog()
        }
        super.onResume()
    }

    public override fun onDestroy() {
        TextureFactory.removeListener()
        bw?.let { blockView ->
            blockView.destroyWorld()
            bw = null
            TextureFactory.deleteAllTextures()
        }
        RegionFileCache.clear()
        System.runFinalization()
        Runtime.getRuntime().gc()
        super.onDestroy()
    }

    override fun finish() {
        world?.let { world ->
            if (world.isNewGame) {
                world.dir.deleteRecursively()
            }
        }
        super.finish()
    }

    private fun showResumeDialog() {
        if (resumeDialog == null) {
            resumeDialog = ProgressDialog(this).apply {
                setTitle("Please wait")
                setMessage("Resuming")
            }
        }
        runOnUiThread { resumeDialog?.show() }
    }

    private fun showProgressDialog() {
        if (loadingProgressDialog == null) {
            loadingProgressDialog = object : CustomProgressDialog(this) {
                override fun onBackPressed() {
                    world?.setCancel(true)
                    finish()
                }

                override fun buttonClick() {
                    world?.setCancel(true)
                    finish()
                }
            }
        }
        loadingProgressDialog!!.show()
    }

    fun dismissProgressDialog() {
        loadingProgressDialog!!.dismiss()
    }

    fun dismissResumeDialog() {
        resumeDialog?.dismiss()
    }

    fun dismissAllLoadingDialogs() {
        loadingProgressDialog?.let { dialog ->
            if (dialog.isShowing) {
                dialog.dismiss()
            }
        }
        resumeDialog?.let { dialog ->
            if (dialog.isShowing) {
                dialog.dismiss()
            }
        }
    }

    inner class TagLoad internal constructor(
        dat: File?,
        val dir: File,
        val isNewGame: Boolean,
        val gameMode: WorldGenerator.Mode?
    ) : TagLoader(dat) {
        override fun complete() {
            runOnUiThread(TagLoaderRunnable())
        }

        inner class TagLoaderRunnable : Runnable {
            override fun run() {
                val mapType: Int
                if (resource == null) {
                    showToast(
                        "Could not load world level.dat ${exception.javaClass.simpleName}: ${exception.message}",
                        true
                    )
                    finish()
                    return
                }
                try {
                    val time = resource.findTagByName(WorldGenerator.LAST_PLAYED)
                    GameTime.initTime((time.value as Long))
                    val playerTag = resource.findTagByName("Player")
                    val pos = playerTag.findTagByName("Pos")
                    val tl = pos.value as Array<Tag>
                    val p = Vector3f()
                    p.x = (tl[0].value as Double).toFloat()
                    p.y = (tl[1].value as Double).toFloat()
                    p.z = (tl[2].value as Double).toFloat()
                    val rotaionTag = playerTag.findTagByName("Rotation")
                    val rotation = Vector2f()
                    if (rotaionTag != null) {
                        val tl2 = rotaionTag.value as Array<Tag>
                        rotation.x = (tl2[0].value as Float)
                        rotation.y = (tl2[1].value as Float)
                        if (rotation.y > 1.5707964f || rotation.y < -1.5707964f) {
                            rotation.y = 0.0f
                        }
                    }
                    val mapTypeTag = resource.findTagByName(WorldGenerator.MAP_TYPE)
                    mapType = if (mapTypeTag != null) {
                        mapTypeTag.value as Int
                    } else {
                        -1
                    }
                    world = object : World(dir, p, resource) {
                        override fun isLoadingDialogVisible(): Boolean {
                            return loadingProgressDialog?.isShowing ?: false
                        }

                        override fun incLoadingProgressStatus(diff: Int) {
                            loadingProgressDialog?.incrementProgressBy(diff)
                        }

                        override fun setLoadingProgressStatus(progress: Int, max: Int) {
                            loadingProgressDialog?.apply {
                                this.max = max
                                this.progress = progress
                            }
                        }

                        override fun showGameMenu() {
                            runOnUiThread { showGameMenuDialog() }
                        }

                        override fun dismissLoadingDialog() {
                            loadingProgressDialog?.dismiss()
                        }

                        override fun dismissLoadingDialogAndWait() {
                            loadingProgressDialog?.let { dialog ->
                                if (dialog.isShowing) {
                                    val latch = CountDownLatch(1)
                                    dialog.setOnDismissListener { latch.countDown() }
                                    dialog.dismiss()
                                    try {
                                        latch.await(2L, TimeUnit.SECONDS)
                                    } catch (e: InterruptedException) {
                                        e.printStackTrace()
                                    }
                                }
                            }
                        }

                        override fun showChat() {
                            runOnUiThread { showChatDialog() }
                        }

                        override fun showReportAbuse() {
                            runOnUiThread {
                                val reportAbuseDialog = ReportAbuseDialog(this@WorldCraftActivity)
                                reportAbuseDialog.show()
                            }
                        }

                        override fun showDeathMenu(player: Player) {
                            runOnUiThread {
                                val dialog = deathMenuDialog
                                if (dialog == null || !dialog.isVisible) {
                                    deathMenuDialog = DeathMenuDialog(
                                        this@WorldCraftActivity,
                                        player
                                    ).apply { show() }
                                }
                            }
                        }
                    }.also { world ->
                        world.setNewGame(isNewGame)
                        world.mapType = mapType
                        (loadingProgressDialog as? CustomProgressDialog)?.updateMax(
                            World.getLoadingLimit(
                                isNewGame
                            )
                        )
                        bw = BlockView(world).also { blockView ->
                            if (isMultiplayerMode) {
                                Multiplayer.instance.blockView = blockView
                            }
                            blockView.setCamRotation(rotation)
                            blockView.cam.invert = Persistence.getInstance().isInvertY
                            setGameMode(if (isMultiplayerMode || WorldGenerator.Mode.SURVIVAL != gameMode) 1 else 0)
                            if (isSurvivalMode && !isMultiplayerMode) {
                                world.initSunLight()
                            }
                            initFog()
                            val game = Game(this@WorldCraftActivity, null, blockView)
                            start(game)
                        }
                    }

                } catch (e: Exception) {
                    showToast("Problem parsing level.dat - Maybe a corrupt file?", true)
                    Log.e(Game.RUGL_TAG, "Level.dat corrupted?", e)
                    finish()
                }
            }

            private fun initFog() {
                var fogDistance = Persistence.getInstance().fogDistance
                if (fogDistance < 0.0f) {
                    fogDistance = 0.0f
                }
                BlockFactory.state.fog.start = fogDistance
                BlockFactory.state.fog.end = 10.0f + fogDistance
            }
        }
    }
}